### PR TITLE
fix(config): Use newer protractor

### LINKED ISSUE #1247: Issue running tests, selenium-webdriver
I am running ubuntu 15.10 but when doing a fresh clone and npm install I am getting:

> [launcher] Process exited with error code 1
> /mean/node_modules/selenium-webdriver/lib/goog/async/nexttick.js:41
>   goog.global.setTimeout(function() { throw exception; }, 0);
>                                       ^
> 
> Error: Server terminated early with status 1
>     at Error (native)
>     at /mean/node_modules/selenium-webdriver/remote/index.js:204:18
>     at promise.ControlFlow.runInFrame_ (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:1857:20)
>     at goog.defineClass.notify (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:2448:25)
>     at promise.Promise.notify_ (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:564:12)
>     at Array.forEach (native)
>     at promise.Promise.notifyAll_ (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:553:15)
>     at goog.async.run.processWorkQueue (/mean/node_modules/selenium-webdriver/lib/goog/async/run.js:130:15)
>     at process._tickCallback (node.js:406:9)
> E2E Tests failed

The tests run fine on a recent fork I created:
https://github.com/trendzetter/mean

### PR DESCRIPTION
**Pull request by trendzetter**
Fixes issue with 2.47 of selenium webdriver
reference angular/protractor#2638
> [launcher] Process exited with error code 1
> /mean/node_modules/selenium-webdriver/lib/goog/async/nexttick.js:41
>   goog.global.setTimeout(function() { throw exception; }, 0);
>                                       ^
> 
> Error: Server terminated early with status 1
>     at Error (native)
>     at /mean/node_modules/selenium-webdriver/remote/index.js:204:18
>     at promise.ControlFlow.runInFrame_ (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:1857:20)
>     at goog.defineClass.notify (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:2448:25)
>     at promise.Promise.notify_ (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:564:12)
>     at Array.forEach (native)
>     at promise.Promise.notifyAll_ (/mean/node_modules/selenium-webdriver/lib/goog/../webdriver/promise.js:553:15)
>     at goog.async.run.processWorkQueue (/mean/node_modules/selenium-webdriver/lib/goog/async/run.js:130:15)
>     at process._tickCallback (node.js:406:9)
> E2E Tests failed

Fixes #1247

### CODE REVIEW

**Comment by mleanos** [comment]
@trendzetter Can you summarize what this is fixing? What types of errors can happen with the tests?

**Comment by mleanos** [comment]
@jloveland Didn't we encounter an issue with using `directConnect` when working on adding Protractor support? 

@trendzetter Will this have any adverse effect on anything else? Will this affect the settings in `.travis.yml`?

**Comment by trendzetter** [comment]
@mleanos I don't know. It just doesn't work without this fix for me and a lot of other people it appears. I have added the error in the summery. For more background please check the reference.

**Comment by trendzetter** [comment]
@mleanos Alternatively I found that adding "protractor": "~3.1.1", to packages.json also resolves the issue. Do you prefer the alternative solution?

**Comment by trendzetter** [comment]
├─┬ gulp-protractor@1.0.0
│ └─┬ protractor@2.5.1

**Comment by trendzetter** [comment]
@lirantal Might fix #1234

**Comment by lirantal** [comment]
yeah that's what I've been wondering about ever since I saw this PR

**Comment by trendzetter** [comment]
adding the line and doing 

> grunt test

Should relieve you of uncertainty.

**Comment by trendzetter** [comment]
> $ node --version
> v5.7.0

**Comment by mleanos** [comment]
I was testing this locally earlier and was encountering other Protractor errors. I'll try upgrading Protractor to your sugesstion. 

I think we want to avoid using `directConnect`. Correct me if I'm wrong. Using that option by-passes using the built in Selenium driver, which would cause issues with our Travis CI builds. 

If the package upgrade works, then let's do that.

**Comment by trendzetter** [comment]
I'm not sure about the bypassing selenium but I do see it executing and displaying chrome.

**Comment by mleanos** [comment]
@trendzetter I believe using `directConnect` will just use the driver of the system's installed version of Chrome, instead of the built-in Selenium server & Chrome driver.

I was having a lot of issues with the client-side tests (including E2E). I did `npm cache clean` & `npm install` a few times. The postinstall kept failing when installing the Bower dependencies. I realized that the correct version of Angular couldn't be resolved. And since the 
```javascript
"resolutions": {
    "angular": "~1.3"
  }
```
was removed when we upgraded to Angular 1.5, I think we're going to have issues. 

Can you try this for me..
1) Remove the `directConnect` option from the Protractor
2) `npm cache clean`
3) `npm install`

If you get an error when the Bower dependencies are installed, try `bower install` and it should ask you to choose a version of Angular to resolve to.

Now check if you're still having issues with Protractor by running the E2E tests.

I'm wondering if we should add
```javascript
"resolutions": {
    "angular": "~1.5"
  }
```
to the package.json

After following these steps I'm not having any issues with the tests at all.

**Comment by trendzetter** [comment]
The resolution in bower.json was needed because previously you used a fixed old file upload dependency. The option directConnect (in protractor.conf.js used at runtime, not install) doesn't have any influence on dependency resolving, it doesn't on my system. I started from a clean clone and cleaning the installation and reinstalling doesn't ask me to resolve dependencies.
Anyway, I see indeed that the end result is selenium is not use rather Chrome which seems a good thing to me but might have issues for other people. 
I will update my PR to add protractor to packages.json instead.

**Comment by mleanos** [comment]
@trendzetter I understand that the `directConnect` option isn't relevant to installing the packages. I was thinking that your local problems could have been related to the Angular version not getting resolved. I was having quite a bit of problems with all the client-side tests, and once I fixed the issue with the Angular version, everything worked perfectly for me. 

I'm still curious if there's some issue that's local to your environment. What is your OS, & NPM version? 

I'm on Windows 7
`node --version v4.3.1`
`npm --version 2.4.12`

@lirantal Are you still having issue with Protractor? If so, did either the `directConnect` option, or installing the Protractor package help?

**Comment by lirantal** [comment]
@mleanos haven't had time to try again the protractor tests.
can you give me exact instructions on how to re-install it / updating the options to pass to it?

**Comment by trendzetter** [comment]
@lirantal If you pull this PR your issue should be solved. No need to pass options I think.

**Comment by trendzetter** [comment]
@mleanos If you are seeing new issues with bower resolution following pr #1207 please create an issue adding the errors. The suggestion that the hard coded dependency resolution isn't needed was made by @ilanbiala and I tend to agree. If you have to manually resolve the dependencies please post the output in the issue. It should be clear which package is blocking automatic resolving from the questions and options.

> $ node -v
> v5.7.1
> $ npm -v
> 3.6.0

Selenium 2.47 has an issue. Now it is installing 2.48 and OK. My guess is you might have issues with these newer versions of protractor or selenium on the oldest versions of node. In the reference in the summary they mention that this happens only in about 50% of the setups but it occurrence is unpredictable.

**Comment by mleanos** [comment]
@lirantal A couple quick steps would be to delete the bower packages in *public/lib*, and then do `bower install`. For me, it couldn't resolve which version of Angular to use, so it presented me with an option. I chose Angular 1.5. Then run the client-side tests to see if your issue is resolved.

If that doesn't help, then another more tedious process would be to delete everything from *node_modules*, do `npm cache clean` & then `npm install`; this is where I noticed the issue with the Angular resolution. Since the post-install script does `bower install --config.interactive=false`, it won't prompt you for choosing what version of Angular to resolve to. So at that time you can perform the quick steps I just described above to re-install the Bower packages & resolve the Angular dependency. 

@trendzetter The reason I brought up the Bower Angular resolution issue, is because I'm trying to determine what exact issues you & Liran are having with the E2E tests, and what the cause might be. I had thought I might be experiencing the same issues as yourself, but it turned out to be the Angular resolution issue.

My goal is determine if there's indeed a reason to add Protractor as a dependency, if others aren't experiencing this issue. Can you elaborate on your comment about Selenium 2.48 being "OK"? Does that mean your issues went away, without the fix in this PR?

**Comment by jloveland** [comment]
Similar to @mleanos, I just pulled down master, ran `npm install` and got `ECONFLICT Unable to find suitable version for angular`. I resolved this by bower installing angularjs 1.5.0. Then I was able to run all the tests using **grunt** without issues. 
screen shot 2016-03-05 at 10 03 46 pm

When running with **gulp** the tests fail.
gulp fail e2e
Seemingly it's a gulp plugin or configuration issue

**Comment by mleanos** [comment]
@jloveland I was actually testing with Gulp, and it's working fine for me. It could be a lingering package issue. Have you tried `npm cache clean` after blowing away *node_modules*? Then the `npm install`.

**Comment by jloveland** [comment]
@mleanos you are right! it works now that I cleaned the cache and installed. I need to add that to my standard practice..

**Comment by mleanos** [comment]
@jloveland Can you share what your environment is? I'm really trying to understand if this is something that we do indeed need to fix, and if the proposed fix is our best option.

@trendzetter Have you tried the whole `npm cache clean` & `npm install`, after blowing away *node_modules*? I still have a hunch that it's an environment issue.

**Comment by jloveland** [comment]
```
$ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.11.3
BuildVersion:	15D21

$ node -v
v0.12.7

$ mongod --version
db version v3.2.3
git version: b326ba837cf6f49d65c2f85e1b70f6f31ece7937
OpenSSL version: OpenSSL 1.0.2g  1 Mar 2016
allocator: system
modules: none
build environment:
    distarch: x86_64
    target_arch: x86_64
```

**Comment by trendzetter** [comment]
@mleanos Can you open an issue for bower asking to resolve dependencies manually since #1207 ? I am willing to look into it if you post the output of just before you manually choose angular 1.5. I am asking you since it's not asking me on a fresh clone so I will have to rely on what you provide.
Concerning this issue here I can assure you that my clone is fresh and I did try deleting of node_modules and clear npm cache etc.. 
I think it is something we can provide a workaround for some external bug, obviously the issue is not with meanjs, we just avoid some buggy version throwing errors in certain combinations (see reference).

**Comment by mleanos** [comment]
@trendzetter Earlier today, I was doing some additional testing with this. I got the same exact error that you originally reported in this PR. I then removed `gulp-protractor`, and then re-installed it. The problem went away for me & the E2E tests ran successfully. 

I would like to avoid adding an additional dependency, and to not use the `directConnect` option. I'm reluctant to merge this fix if others aren't reporting issues, that can't be solved with the steps that myself, and @jloveland have confirmed.

**Comment by mleanos** [comment]
The main point I want to make is that there are a lot of environmental factors that come into play when dealing with node packages. 

I've had a rough go at it ever since I started using Node & NPM. I constantly get random errors when installing packages through NPM. It's like pulling teeth, and then getting smacked in the back of the head for me :) I have been resolved to chalking it up to my environment; a mix of potential Windows compatibility issues, and perhaps some flaw in my setup. However, I'm always able to overcome it with a little bit of patience with re-installing packages & using tools like `npm cache clean`. I'm just trying to rule it out, as some sort of environmental issue on your side.

I don't want to open an issue about the Angular bower dependency resolution. For one thing, I can't reproduce the issue any longer locally.

**Comment by trendzetter** [comment]
I tried several times deleting the node_modules but now it works on a fresh clone. And I flushed then npm cache but maybe I did it in the wrong order. Sorry for the fuss, still willing to look into the bower dependency resolution.

**Comment by mleanos** [comment]
Ok. I'm glad you were able to resolve the issue. I know it was a frustrating back & forth. 

As for the Bower dependency resolution, I just can't reproduce :) I think we're probably fine on that end.

Thanks for your work with this, and perhaps it could be helpful for others coming along with similar issues.

**Comment by lirantal** [comment]
@mleanos I'm still having issues with running karma as well as protractor now.
Due to my effort with a docker container I'll try to see how it works there rather than locally.
